api/exec/compose_stack.go

```diff
@@ -148,28 +151,46 @@ func createEnvFile(stack *portainer.Stack) (string, error) {
 	}
 	defer envfile.Close()
 
-	copyDefaultEnvFile(stack, envfile)
+	// Copy from default .env file
+	defaultEnvPath := path.Join(stack.ProjectPath, path.Dir(stack.EntryPoint), ".env")
+	if err = copyDefaultEnvFileIfExists(envfile, defaultEnvPath); err != nil {
+		return "", err
+	}
 
-	for _, v := range stack.Env {
-		envfile.WriteString(fmt.Sprintf("%s=%s\n", v.Name, v.Value))
+	// Copy from stack env vars
+	if err = copyConfigEnvVarsIfExists(envfile, stack.Env); err != nil {
+		return "", err
 	}
 
 	return "stack.env", nil
 }
 
 // copyDefaultEnvFile copies the default .env file if it exists to the provided writer
-func copyDefaultEnvFile(stack *portainer.Stack, w io.Writer) {
-	defaultEnvFile, err := os.Open(path.Join(path.Join(stack.ProjectPath, path.Dir(stack.EntryPoint)), ".env"))
+func copyDefaultEnvFileIfExists(w io.Writer, defaultEnvFilePath string) error {
```

Comment:
```suggestion
func copyDefaultEnvFile(w io.Writer, defaultEnvFilePath string) error {
```

I feel like `IfExists` doesn't add any value. it copied the file into the writer

Reply:
When there is no .env file, we skip the copy action